transports/websocket/src/framed.rs

```diff
@@ -309,6 +289,74 @@ where
         Ok(Box::pin(future))
     }
 
+    fn is_tcp_addr(&self, addr: &Multiaddr) -> bool {
+        let mut addr_iter = addr.iter();
+        matches!(
+            addr_iter.next(),
+            Some(Protocol::Ip4(_)) | Some(Protocol::Ip6(_))
+        ) && matches!(addr_iter.next(), Some(Protocol::Tcp(_)))
+            && match addr_iter.next() {
+                Some(Protocol::Wss(_)) if self.tls_config.server.is_none() => {
+                    debug!("/wss address but TLS server support is not configured");
+                    false
+                }
+                Some(Protocol::Wss(_)) | Some(Protocol::Ws(_)) => true,
+                _ => false,
+            }
+    }
+
+    fn insert_proto(&mut self, inner_addr: Multiaddr, proto: Protocol<'static>) {
+        let mut addr_iter = inner_addr.iter();
+        let ip_proto = addr_iter.next().expect("Address is valid.");
```

Comment:
These `expect` calls are safe due to the previous call to `is_tcp_addr`, correct? Given that the two methods are called in sequence, how about combining them into one method, e.g. `validate_addr_and_insert_proto`? That would remove the need for an `expect` call, right?

I would like to be a bit careful, given that a remote could potentially give us a malformed address and thus panic the main event loop in case miss an edge case here.